Sources/NIOFoundationCompat/Codable+ByteBuffer.swift

```diff
@@ -0,0 +1,59 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2019 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+import Foundation
+
+extension JSONDecoder {
+    /// Returns a value of the type you specify, decoded from a JSON object inside a `ByteBuffer`.
+    ///
+    /// If the `ByteBuffer` does not contain valid JSON, this method throws the
+    /// `DecodingError.dataCorrupted(_:)` error. If a value within the JSON
+    /// fails to decode, this method throws the corresponding error.
+    ///
+    /// - parameters:
+    ///     - type: The type of the value to decode from the supplied JSON object.
+    ///     - buffer: The `ByteBuffer` that contains JSON object to decode.
+    /// - returns: The decoded object.
+    public func decode<T: Decodable>(_ type: T.Type, from buffer: ByteBuffer) throws -> T {
```

Comment:
Since `buffer` isn't `inout` the bytes in the callers buffer won't be read: we should call this out in the documentation.

Maybe we can add another API to `ByteBuffer` to read a `Decodable` which _does_ consume the bytes: this would  be more inline with the other `read` APIs.

Reply:
great idea, done.